packages/firestore/src/local/memory_mutation_queue.ts

```diff
@@ -257,17 +257,18 @@ export class MemoryMutationQueue implements MutationQueue {
     transaction: PersistenceTransaction,
     documentKeys: DocumentKeySet
   ): PersistencePromise<MutationBatch[]> {
-    if (documentKeys.isEmpty()) {
-      return PersistencePromise.resolve([]);
-    }
-    const start = new DocReference(documentKeys.first(), 0);
-    const end = new DocReference(documentKeys.last(), Number.POSITIVE_INFINITY);
     let uniqueBatchIDs = new SortedSet<number>(primitiveComparator);
 
-    this.batchesByDocumentKey.forEachInRange([start, end], ref => {
-      if (documentKeys.has(ref.key)) {
+    documentKeys.forEach(documentKey => {
+      const start = new DocReference(documentKey, 0);
+      const end = new DocReference(documentKey, Number.POSITIVE_INFINITY);
+      this.batchesByDocumentKey.forEachInRange([start, end], ref => {
+        if (!documentKey.isEqual(ref.key)) {
+          return;
```

Comment:
nit: Is it possible to hit this?  Seems like our range should constrain us to `documentKey`.

Reply:
I'm not sure. This function is derived from `getAllMutationBatchesAffectingDocumentKey`, which [asserts](https://github.com/firebase/firebase-js-sdk/blob/87e0b06bdf5fdd78ff57fe2b7e47207b33542bdc/packages/firestore/src/local/memory_mutation_queue.ts#L241-L244) this equality, this led me to believe this case is not impossible. I can remove it altogether or replace with an assertion.

Reply:
Asserts should generally mean the case is impossible.  They exist to document our expectations and/or catch false expectations as early as possible to make finding / fixing bugs later. So I think an assert here makes perfect sense.

Fine print: We're probably not 100% rigorous about asserts always being impossible.  E.g. there may be cases where we use asserts for cases that /are/ possible but represent rare error conditions that we don't feel compelled to provide proper error handling for.

Reply:
Replicated the assertion.